dosbox-staging/dosbox-staging
# Implement read-only mounting + make Y drive read-only by default

# Description

I mostly stuck to my plan here:  https://github.com/dosbox-staging/dosbox-staging/issues/2602#issuecomment-2134255705

1. Implement a `readonly` flag for `localDrive`.  This already exists for `fatDrive`.  If `readonly` is set, throw a `DOSERR_ACCESS_DENIED` on file writes, creation, and deletions.
2. Implement a `-ro` flag for the `mount` command.  This already exists for `imgmount` but not for `mount`.
3. Add a `readonly` config setting to `[drive]`.  This will simply pass the `-ro` flag to `mount`.
4. Add `readonly = true` to `contrib/resources/drives/y.conf` so it is read-only by default.  Users can of course change this if they want.

The first point just took a lot more commits than I originally thought :smile: 

2, 3, and 4, were very simple.  A single small commit each (last 3 commits of this PR).

The `DOS_Drive` classes could really use some refactoring in some more places even but I did a fair amount of cleanup here.  Ideally, more of the logic would be handled in the generic `DOS_` functions in `dos_files.cpp` regarding what DOS errors to throw and such.

## Related issues

Fixes #2602


# Manual testing

Removing, deleting and creating files is not allowed when `-ro` is set by the mount command, as expected.

The Y drive gets mounted read-only as expected.  May need to do a clean build so it copies the new `y.conf` to the build directory.

Just to see what would happen, I mounted my C drive read-only and tested Doom, Space Quest 1, Space Quest 3.  These games work, only throwing an error when attempting to save the game.  Tyrian does not launch, and this gets logged to console.

```
2024-05-29 16:38:21.983 | FILESYSTEM: protected from modification: palette.dat
2024-05-29 16:38:21.998 | FILESYSTEM: protected from modification: tyrian.cfg
```

It is expected that some games won't like it if they can't write to the hard drive.


# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:
